### PR TITLE
Improve email regexp on edge cases

### LINKED ISSUE #10600: Email parsing slowdown on edgecases
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I found that current pydantic-level regexp is not optimal and may cause unexpected slowdowns on some edgecases. Email strings like `<        ` (+many spaces). 

POC below shows it. You can run it  with command like `/usr/bin/time python pydantic-poc.py 500`.
The third value in a line should be near constant which shows that approximate complexity of one operation is ~ O(N^3)

I believe it is not a security issue because it runs fast enough even for largest possible sequences however.

### Example Code

```Python
import sys

from pydantic import BaseModel, EmailStr, ValidationError
import time


class TestModel(BaseModel):
    email: EmailStr


def main(max_ind: int):
    for ind in range(1, max_ind):
        begin_at = time.time()
        try:
            TestModel.parse_obj({"email": "<" + " " * ind})
        except ValidationError:
            pass
        delta = time.time() - begin_at
        print(
            f"{ind} takes: {delta:0.6f}, delta**1/3 / ind is {delta**(1/3) / ind:0.6f}"
        )
    return


if __name__ == "__main__":
    main(int(sys.argv[1]))
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.9.2
        pydantic-core version: 2.23.4
          pydantic-core build: profile=release pgo=false
                 install path: /home/alex/Projects/re-conference/venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]
                     platform: Linux-6.8.0-45-generic-x86_64-with-glibc2.39
             related packages: typing_extensions-4.12.2
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by AlekseyLobanov**
- Drastically improves performance on cases like `"<" + " " * N`
- Last spaces are not needed anyway because this group is stripped later. Also spaces will be caught by `.` anyway.

## Change Summary

I found that one single change in email regexp solves slowdowns on special invalid email strings. See related issue for details

## Related issue number

Fixes #10600 

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by AlekseyLobanov** [comment]
please review

**Comment by AlekseyLobanov** [comment]
How performance changes?
I use my own POC in #10600 and run it as `/usr/bin/time python pydantic-poc.py 500`

- **Before:** 5.35user 0.01system 0:05.37elapsed 99%CPU (0avgtext+0avgdata 36840maxresident)k
- **After:**0.20user 0.01system 0:00.22elapsed 99%CPU (0avgtext+0avgdata 37004maxresident)k

About **25x** speed improvement.

File: pydantic/networks.py [hunk 1]
```diff
@@ -722,7 +722,7 @@ def _build_pretty_email_regex() -> re.Pattern[str]:
     name_chars = r'[\w!#$%&\'*+\-/=?^_`{|}~]'
     unquoted_name_group = rf'((?:{name_chars}+\s+)*{name_chars}+)'
     quoted_name_group = r'"((?:[^"]|\")+)"'
-    email_group = r'<\s*(.+)\s*>'
+    email_group = r'<\s*(.+)>'
```
Comment (Viicos) [inline pydantic/networks.py:725]:
```suggestion
    email_group = r'<(.+)>'
```

The `.strip()` call also removes leading whitespaces.

**Review by Viicos — COMMENTED** [review]
Thanks, this looks reasonable but to be extra careful we'll wait for other reviews as well.

Could you add the following to the `test_address_valid` test?:

```python
        ('Samuel Colvin < s@muelcolvin.com>', 'Samuel Colvin', 's@muelcolvin.com'),
        ('Samuel Colvin <s@muelcolvin.com >', 'Samuel Colvin', 's@muelcolvin.com'),
        ('Samuel Colvin < s@muelcolvin.com >', 'Samuel Colvin', 's@muelcolvin.com'),
```

**Comment by AlekseyLobanov** [comment]
> Thanks, this looks reasonable but to be extra careful we'll wait for other reviews as well.

According to [Wikipedia](https://en.wikipedia.org/wiki/ReDoS) it is one of the valid DoS attack vectors. And at least some of known to me rate limiters will work only _after_ validation step. 

> Could you add the following to the test_address_valid test?

I think that existing tests are already covering this edge cases (spaces before/after the group). Should I still add yours?

```python
        ('foo BAR <foobar@example.com >', 'foo BAR', 'foobar@example.com'),
        ('FOO bar   <foobar@example.com> ', 'FOO bar', 'foobar@example.com'),
        ('Whatever < foobar@example.com>', 'Whatever', 'foobar@example.com'),
```

**Comment by sydney-runkle** [comment]
Yep let's add those extra tests and fix the lints, but otherwise LGTM.

**Comment by Viicos** [comment]
> According to [Wikipedia](https://en.wikipedia.org/wiki/ReDoS) it is one of the valid DoS attack vectors. And at least some of known to me rate limiters will work only _after_ validation step.

I agree, just wanted to be careful as changing regex can be a source of breaking changes.

> I think that existing tests are already covering this edge cases (spaces before/after the group). Should I still add yours?

Missed these ones, then maybe only add these ones after it:

```python
        ('Whatever <foobar@example.com >', 'Whatever', 'foobar@example.com'),
        ('Whatever < foobar@example.com >', 'Whatever', 'foobar@example.com'),
```

**Review by sydney-runkle — APPROVED** [review]
Nice, thanks for the help here! We appreciate the thorough explanations / refs :).
